Add tools and Debian packaging for parsing dlopen ELF notes

bluca opened the pull request:
Spec and implementation at https://github.com/systemd/systemd/pull/32234

keszybz commented on dlopen-notes.py:
```diff
@@ -0,0 +1,62 @@
+
+if __name__ == '__main__':
+    if len(sys.argv) < 2:
+        print('Usage: %s <elf-file> [<elf-file> ...]' % sys.argv[0])
+        sys.exit(1)
```
`%` should not be used in new code. Really.

```python
sys.exit(f'Usage: {sys.argv[0]} …')
```

keszybz commented on dlopen-notes.py:
```diff
@@ -0,0 +1,61 @@
+        sys.exit(f'Usage: {sys.argv[0]} <elf-file>...')
+
+    sonames = {}
+    for i in range(1, len(sys.argv)):
+        sonames.update(dlopen_notes_parse(sys.argv[i]))
```
```
for filename in sys.argv[1:]:
    sonames.update(dlopen_notes_parse(filename))
```

keszybz commented on dlopen-notes.py:
```diff
@@ -0,0 +1,61 @@
+def dlopen_notes_parse(elf):
+    sonames = {}
+
+    with open(elf, 'rb') as f:
```
`with` is actually not useful here, since (read-only) open files are just another resource managed automatically by python and the system.

keszybz commented on dlopen-notes.py:
```diff
@@ -0,0 +1,61 @@
+
+                    for e in j:
+                        soname = e.get('soname')
+                        priority = e.get('priority')
+
+                        if not soname:
+                            continue
+
+                        if not priority:
+                            priority = 'recommended'
```
`priority = e.get('priority', 'recommended')` ?

bluca commented:
> I pushed another commit to rework the output to have more postprocessing.

Merged into the first commit with Co-authored-by - also added CI, and added back a way to print a "soname priority" list, that I need for the Debian tool, behind a new option

bluca commented:
I have no idea why it errors out in the CI, the string seems fine... also works locally, any idea?

behrmann commented:
The bit
```
[{"feature":"fido2","description":"Support fido2 for encryption and authentication.","priority":"required","soname":["libfido2.so.1"]}]
```
in the output makes me think that the else path here
```python
            try:
                if not isinstance(note_desc, str):
                    text = note_desc.decode('utf-8').rstrip('\0')
                else:
                    text = note_desc
                    print(text)
            except UnicodeDecodeError as e:
                raise ValueError(f'{filename}: Invalid UTF-8 in .note.dlopen n_desc') from e
```
is taken, but when I copy the above JSON into Emacs, it has a trailing null byte, which gives me the same error
> json.decoder.JSONDecodeError: Extra data: line 1 column 136 (char 135)

Yes, unfortunately strings in Python can have null bytes, so stripping that might be what's needed.

bluca commented:
Indeed stripping that side of the if fixes the issue, thanks

keszybz commented on dlopen-notes.py:
```diff
@@ -0,0 +1,163 @@
+
+            yield from j
+
+def group_by_soname(notes):
+    sonames = {}
+
+    for note in notes:
+        for element in note:
+            priority = element.get('priority', 'recommended')
+            for soname in element['soname']:
+                sonames[soname] = priority
+
+    return sonames
```
```python
def dictify(f):
    def wrap(*args, **kwargs):
        return dict(f(*args, **kwargs))
    return functools.update_wrapper(wrap, f)

@dictify
def group_by_soname(notes):
    for note in notes:
        for element in note:
            priority = element.get('priority', 'recommended')
            for soname in element['soname']:
                yield soname, priority
```
?

keszybz commented on test/test.py:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: CC0-1.0
+
+import unittest
```
Let's not use unittest, it's an unproductive java-inspired monstrosity.

Just drop the `class` boilerplate and pytest will pick up the test with any issue.
